deltacat/compute/compactor_v2/utils/primary_key_index.py

```diff
@@ -167,6 +167,40 @@ def group_by_pk_hash_bucket(
     return result
 
 
+def _sliced_string_cast(array: pa.ChunkedArray) -> pa.ChunkedArray:
+    """performs slicing of a pyarrow array prior casting to a string.
+    This prevents a pyarrow from allocating too large of an array causing a failure.
+    """
+    dtype = array.type
+    MAX_BYTES = 2147483646
+    max_str_len = None
+    if pa.types.is_integer(dtype):
+        max_str_len = 21  # -INT_MAX
+    elif pa.types.is_floating(dtype):
+        max_str_len = 24
+    elif pa.types.is_decimal128(dtype):
+        max_str_len = 39
+    elif pa.types.is_decimal256(dtype):
+        max_str_len = 77
```

Comment:
A couple minor asks:
1. I think both signed integers (-2^63) and unsigned integers (2^64) have a max length of 20 bytes. Am I missing something here that led to the conclusion of 21 bytes?
2. Can we add some comments giving a bit more context to each figure here... maybe something like the following:
```suggestion
        max_str_len = 21  # UINT64_MAX (2^64) / -INT64_MAX(-2^63)
    elif pa.types.is_floating(dtype):
        max_str_len = 24  # ??? TODO
    elif pa.types.is_decimal128(dtype):
        max_str_len = 39. # ??? TODO
    elif pa.types.is_decimal256(dtype):
        max_str_len = 77 # ??? TODO
```
Or, to provide this clarity implicitly, what do you think about having constants where we define the pyarrow type that is assumed to have the max string length, cast it to a string, and count the digits once, for example:
```
PA_UINT64_MAX_STR_BYTES = pa.compute.binary_length(pa.compute.cast(pa.scalar(2**64-1, type=pa.uint64()), pa.string())).as_py()
PA_INT64_MAX_STR_BYTES = pa.compute.binary_length(pa.compute.cast(pa.scalar(-2**63, type=pa.int64()), pa.string())).as_py()
PA_INT_MAX_STR_BYTES = max(PA_UINT64_MAX_STR_BYTES, PA_INT64_MAX_STR_BYTES)
```

Reply:
I mistakenly did `(-2^64)` which led to that. I moved to cached functions that compute the various lengths. (Moved to `deltacat/utils/pyarrow.py`)